src/read-only-handler.ts

```diff
@@ -1,31 +1,60 @@
 import {
-    isUndefined,
     ArrayConcat,
-    ObjectDefineProperty,
-    getOwnPropertyDescriptor,
     getOwnPropertyNames,
     getOwnPropertySymbols,
-    hasOwnProperty,
+    getPrototypeOf,
+    unwrap,
 } from './shared';
 
 import {
     ReactiveMembrane,
     ReactiveMembraneShadowTarget,
-    wrapDescriptor,
+    getOwnPropertyDescriptorMembraneTrap,
+    isExtensibleMembraneTrap,
+    MembraneProxyHandler,
 } from './reactive-membrane';
 
-function wrapReadOnlyValue(membrane: ReactiveMembrane, value: any): any {
-    return membrane.valueIsObservable(value) ? membrane.getReadOnlyProxy(value) : value;
-}
+const getterMap = new WeakMap<() => any, () => any>();
+const setterMap = new WeakMap<(v: any) => void, (v: any) => void>();
 
-export class ReadOnlyHandler {
-    private originalTarget: any;
-    private membrane: ReactiveMembrane;
+export class ReadOnlyHandler implements MembraneProxyHandler {
+    originalTarget: any;
+    membrane: ReactiveMembrane;
 
     constructor(membrane: ReactiveMembrane, value: any) {
         this.originalTarget = value;
         this.membrane = membrane;
     }
+    wrapValue(value: any): any {
+        const { membrane } = this;
+        return membrane.valueIsObservable(value) ? membrane.getReadOnlyProxy(value) : value;
+    }
+    wrapGetter(originalGet: () => any): () => any {
+        if (getterMap.has(originalGet)) {
+            return getterMap.get(originalGet) as () => any;
+        }
+        const handler = this;
+        function get(this: any): any {
+            // invoking the original getter with the original target
+            return handler.wrapValue(originalGet.call(unwrap(this)));
+        };
+        getterMap.set(originalGet, get);
+        return get;
+    }
+    wrapSetter(originalSet: (v: any) => void): (v: any) => void {
+        if (setterMap.has(originalSet)) {
+            return setterMap.get(originalSet) as (v: any) => void;
+        }
+        const handler = this;
+        function set(this: any, v: any) {
```

Comment:
@ravijayaramappa this is probably the biggest difference. In the old code, the setter on read-only handlers was always undefined, but getting it to undefined will imply that we will not be able to unwrap it to the original implementation anymore. Instead, we are producing a setter, but it throws when invoked. That seems to be equivalent in behavior, but it is an observable difference.